webhooks/webhook.go

```diff
@@ -0,0 +1,134 @@
+package webhooks
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	log "github.com/sirupsen/logrus"
+	"golbat/config"
+	"golbat/geo"
+	"io"
+	"net/http"
+	"net/url"
+)
+
+var webhookConfigStringToType = map[string]WebhookType{
+	"gym":           GymDetails,
+	"raid":          Raid,
+	"quest":         Quest,
+	"pokestop":      Pokestop,
+	"invasion":      Invasion,
+	"weather":       Weather,
+	"fort_update":   FortUpdate,
+	"pokemon_iv":    PokemonIV,
+	"pokemon_no_iv": PokemonNoIV,
+}
+
+type webhook struct {
+	url         string
+	areaNames   []geo.AreaName
+	typesWanted []WebhookType
+	httpClient  *http.Client
+}
+
+func (wh *webhook) getPayload(collection webhookCollection) ([]byte, error) {
+	var totalCollection []webhookMessage
+
+	if len(wh.areaNames) == 0 {
+		for _, whType := range wh.typesWanted {
+			totalCollection = append(
+				totalCollection,
+				collection[whType].Messages...,
+			)
+		}
+	} else {
+		for _, whType := range wh.typesWanted {
+			for _, message := range collection[whType].Messages {
+				if geo.AreaMatchWithWildcards(message.Areas, wh.areaNames) {
+					totalCollection = append(totalCollection, message)
+				}
+			}
+		}
+	}
+
+	log.Infof("There are %d webhooks to send to %s", len(totalCollection), wh.url)
+
+	return json.Marshal(totalCollection)
+}
+
+func (wh *webhook) sendCollection(collection webhookCollection) error {
+	payload, err := wh.getPayload(collection)
+	if err != nil {
+		return fmt.Errorf("failed to generate payload: %s", err)
+	}
+
+	req, err := http.NewRequest("POST", wh.url, bytes.NewReader(payload))
+	if err != nil {
+		return fmt.Errorf("failed to create http request to %s: %s", wh.url, err)
+	}
+
+	req.Header.Set("X-Golbat", "hey!")
+	req.Header.Set("Content-Type", "application/json")
+
+	resp, err := wh.httpClient.Do(req)
+	if err != nil {
+		return fmt.Errorf("failed to send webhook to %s: %s", wh.url, err)
+	}
+
+	defer func() {
+		// full body must be read to reuse keep-alive connections.
+		io.Copy(io.Discard, resp.Body)
+		resp.Body.Close()
+	}()
+
+	// TODO: implement retries upon request fails
```

Comment:
Yeah, I'm personally just kind of picky and I have a thing for "no downtime upgrades". If something is poorly performing, that is one thing, and I agree with you for that case. Retries are just a big waste of "work" for no reason for that.

What I had in mind was for the times that you are restarting Poracle, for example. I mean, there's probably a very low chance at this, but what happens if in that one second that Poracle is down... you try to webhook a hundo frigibax and it gets dropped? I would be sad. But then.. there's also a low chance that I would ever know, so I don't know. :)

(FWIW, this is a very similar reason to why I added the graceful shutdown and flushing of webhooks in here before exiting.)

I think there's probably a reasonable way to implement it such that if things are backed up, retries get disabled.
And If it were to be done, I'd make it configurable, of course, and defaulting it to off would also be reasonable.

I'll remove this TODO, though. It's not my project so I probably shouldn't be saying that this one is a TODO.

Reply:
If retries were intelligent somehow - eg tried a few times for up to 30 seconds then stopped retrying all attempts at this destination then it could be fine. 
what I would hate is to get into a situation where a down destination was causing a massive spike of retries and held data on golbats side

Reply:
Agree. That's no good.